backend/amd64/reload.ml

```diff
@@ -146,6 +146,29 @@ method! reload_operation op arg res =
       if !Clflags.pic_code || !Clflags.dlcode || Arch.win64
       then super#reload_operation op arg res
       else (arg, res)
+  | Icsel tst ->
+      (* Last argument and result must the same register.
+         Result must be in register. The last two arguments are used
+         for emitting cmov, the remaining for [Mach.test]. *)
+    (*  CR gyorsh: we already use Array.sub here,
+        so no reason for this convoluted arrangement,
+        using the first two args for cmov would simplify most of the
+        code as it won't need to have [len], it will be able to have indexes
+        directly, but then in Emit we will have to do Array.sub again
+        to call emit_test (unless emit_test takes an index, which is also
+        weird). *)
+      if stackp res.(0)
+      then begin
+        let r = self#makereg res.(0) in
+        let len = Array.length arg in
+        let arg' = Array.copy arg in
+        let test_arg = self#reload_test tst (Array.sub arg 0 (len - 2)) in
+        for i = 0 to len - 2 do
+          arg'.(i) <- test_arg.(i)
+        done;
+        arg'.(len - 1) <- r;
+        (arg', [|r|])
```

Comment:
It looks like we may lose some sharing
between the arguments for the test,
and the last two. Could this be an issue?

Reply:
yes. arguments of both the test and the move parts can change independently.  I'm not sure if it is always worse if sharing is lost, probably. It's not a correctness concern, but definitely worse codegen most of the time with the current regalloc.